man/powerpanel.8
.TH POWERPANEL 8 "Tue Dec 23 2008" "" "Network UPS Tools (NUT)"
.SH NAME
powerpanel \- Driver for PowerPanel Plus compatible UPS equipment
.SH NOTE
This man page only documents the hardware\(hyspecific features of the
powerpanel driver.  For information about the core driver, see
\fBnutupsdrv\fR(8).

.SH SUPPORTED HARDWARE
This driver supports CyberPower BC1200, PR2200 and many other similar
devices, both for the text and binary protocols. The driver will
autodetect which protocol is used.

.SH EXTRA ARGUMENTS
This driver supports the following optional settings in \fBups.conf\fR(5):

.IP "protocol=\fI[text,binary]\fR"
Override the default autodetection of the protocol.

.IP "manufacturer=\fIvalue\fR"
If you don't like the autodetected value, you can override this by setting
it here.

.IP "model=\fIvalue\fR"
Like manufacturer above.

.IP "serial=\fIvalue\fR"
Like manufacturer above.

.IP "ondelay=\fIvalue\fR"
Time to wait before switching on the UPS (1 - 9999 minutes, 0 indefinite).
Only available with the text protocol driver (see \fBSUPPORT STATUS\fR).

.IP "offdelay=\fIvalue\fR"
Time to wait before shutting down the UPS (6 - 600 seconds). Values below 60
seconds will be truncated to 6 seconds intervals, values above 60 seconds to
60 seconds intervals. Only available with the text protocol driver (see
\fBSUPPORT STATUS\fR).

.SH VARIABLES
Depending on the type of your UPS unit, some of the following variables may
be changed. If the driver can't read a variable from the UPS, it will not be
made available.
.TP 12
.B input.transfer.high
writable: high transfer voltage point in V
.TP
.B input.transfer.low
writable: low transfer voltage point in V
.TP
.B battery.charge.low
writable: remaining battery charge percentage for low battery warning
.TP
.B output.voltage.nominal
writable: nominal output voltage in V
.TP
.B ups.start.battery
writable: allow cold start from battery
.PD

.SH COMMANDS
Depending on the type of your UPS unit, some of the following commands may
be available.
.TP 12
.B test.battery.start.quick, test.battery.stop
.TP
.B beeper.enable, beeper.disable, beeper.toggle
.TP
.B shutdown.return, shutdown.reboot, shutdown.stayoff
On many devices, these commands are unreliable, so before using them you
must verify that these work as expected (see \fBSHUTDOWN ISSUES\fR).
.TP
.B shutdown.stop
.PD

.SH SUPPORT STATUS
Vendor support is absent for this driver, so if you need some features that
are currently not available, provide ample documentation on what the driver
should sent to the UPS in order to make this work. If more information
would be available on the binary protocol, it would probably be possible to
make ondelay and offdelay configurable. So far, nobody has taken the time
to investigate what we should tell the UPS to make this work and CyberPower
isn't willing to share this with us.

.SH SHUTDOWN ISSUES
If the \fBshutdown.return\fR command on your UPS doesn't seem to work,
chances are that your UPS is of an older model. Try a couple of different
settings for \fIoffdelay\fR. If no value in the range 6..600 works, your
UPS likely doesn't support this. In order to get the expected behaviour,
it requires \fBshutdown.stayoff\fR (when on battery) and \fBshutdown.reboot\fR
(when on mains). The driver will automatically fallback to these commands if
\fBshutdown.return\fR fails and tries to detect which one should be used when
called with the '\(hyk' option (or through \fBupsdrvctl shutdown\fR).

This isn't bullet proof however and you should be prepared that the
power will either not be shutdown or that it doesn't return when the
power comes back. All models supported by the binary protocol and many
supported through the text protocol are affected by this.

.SH KNOWN PROBLEMS
The CyberPower OP series don't offer direct voltage, charge, frequency
and temperature readings. Instead, they will return a binary value
that needs conversion to the actual value.

The exact conversion needed is unknown at the time of this writing,
hence an estimation was made based om readings from actual devices.
This may (probably will) be off, possibly a lot. Unless you can tell
us the exact mapping between values from the UPS and actual readings,
don't bother to complain. We've done the best we can based on the
limited information available. Remember, a UPS isn't a measuring
instrument.

.SH AUTHORS
Arjen de Korte <arjen@de\(hykorte.org>, Doug Reynolds <mav@wastegate.net>

.SH SEE ALSO

.SS The core driver:
\fBnutupsdrv\fR(8)

.SS Internet resources:
The NUT (Network UPS Tools) home page: http://www.networkupstools.org/


NEWS
If you're upgrading from an earlier version, see the UPGRADING file.

---------------------------------------------------------------------------
Release notes for NUT 2.4.0 - what's new since 2.2.2:

 - preliminary support for Power Distribution Units (PDUs): NUT now support
   PDUs, either natively (ie using NUT snmp-ups driver), or through a binding to
   the Powerman daemon. The list of supported PDUs is already quite long,
   including: Eaton ePDUs (Managed and Monitored), some Aphel models, some
   Raritan PDUs, and the whole list of Powerman supported devices:
   http://powerman.sourceforge.net/supported.html
 
 - support for new devices: the various PDUs cited above, Centralion Micro
   Master, Chloride Desk Power 650, Cyber Power Systems Value 400E/600E/800E
   (USB models), Delta GES602N, the whole Eaton ranges (mostly composed of
   MGE Office Protection Systems and Powerware units), Forza Power Technologies
   SL-1001, HP PowerTrust 2997A, HP R/T 2200 G2, Infosec XP 1000 and XP 500,
   Ippon Back Power Pro (serial and USB), Kebo 1200D/D Series, Liebert PowerSure
   Personal XT, MGE Office Protection Systems Protection Station, Neus 400va and
   600va, Phasak 400VA and 600VA, Plexus 500VA, Powercom Black Knight PRO / King
   PRO and Imperial, PowerKinetics BlackOut Buster, UNITEK Alpha 500, WinPower
   CPM-800.

 - NUT now embeds Python client support in through the PyNUTClient module and
   the NUT-Monitor application. Both are from David Goncalves, and are still
   available from http://www.lestat.st.
   For more information, refer to scripts/python/client/README.

 - the dummy-ups driver now support a "repeater" mode. This allows it to act as
   a NUT client, and to forward data. This can be useful for supervision and
   load sharing purposes.

 - tcp-wrappers support has been added to the upsd server, to grant users access
   by source IP for commands that require to be logged into the server. This
   replaces the previous internal implementation.

 - the nut.conf file has been introduced to standardize startup configuration
   across the various systems.

 - many internal changes to improve maintenability, while lowering the
   maintenance cost (thus allowing developers to focus on what matters: the
   code!). Examples of this are:
   - the USB information automatic extraction to generate the various USB helper
     files,
   - the upsdrv_info_t structure to track more driver information, and remove
     the need for the upsdrv_banner() function
   - common USB code refactoring, as it is done for the serial functions.

 - tons of bugfixes, cleanup and improvements to make NUT stronger than ever!

 For a complete list of changes, please refer to the ChangeLog file.

---------------------------------------------------------------------------
Release notes for NUT 2.2.2 - what's new since 2.2.1:

 - support for new devices: APC BACK-UPS XS LCD, Atlantis Land,
   Mustek Powermust Office 650, Oneac XAU models, Powerware PW5115 and
   PW9120 (USB), Nitram Elite 2005
 
 - Integrated Power Management (NUT HAL integration) has reached a
   major milestone: it is now the most advanced UPS integration into
   Power Management layer known in existing OSs. It has received many
   corrections and improvements, and allows to PowerOff the UPS at the
   end of a power cycle (which is the most important feature, not
   supported on other systems).
   The various files are now installed into the correct location.

 - the usbhid-ups driver has received attention. Most notably, the
   shutdown handling has been reworked, and support for MGE UPS SYSTEMS
   3 phases units has been added.

 - snmp-ups now supports MGE* Environment Sensor (ref 66 846).
   The ambient.temperature reporting has also been fixed for units
   other than APC.

 - the netxml-ups driver has appeared to support MGE* network HTTP/XML
   cards.
 
 - NUT now distributes by default the shared version of libupsclient
   (version 1.0.0), and use this for the provided clients (upsmon, upsc,
   upsrw, upscmd). This is part of an effort to reduce NUT's footprint,
   both on disk and in memory.
 
 - powerpanel has reach a new step toward the replacement of nitram and
   cpsups drivers. The final step is scheduled for NUT 2.4.
 
 - many changes, cleanup and fixes to the NUT core and various drivers.

 For a complete list of changes, please refer to the ChangeLog file.

---------------------------------------------------------------------------
Release notes for NUT 2.2.1 - what's new since 2.2.0:

 - support for new devices: all MGE Office Protection Systems units,
   Advice TopGuard 2000, Belkin F6H375-USB, Dynamix UPS1700D, Effekta RM2000MH,
   Jageson Technology Jasuny USPS, Powercom SMK-1500A and SXL-1500A,
   PowerWalker Line-Interactive VI 400/800 and 600, Powerware 9110,
   UNITEK Alpha 2600, UPSonic CXR1000, some vintage serial APC UPSs.

 - the usbhid-ups driver has been improved, and fixed in many areas, through
   a backport of the development (trunk) version.

 - the udev rules, for Linux hotplug support of the USB UPSs, has been
   updated to support kernel newer than 2.6.22.

 - the megatec and megatec_usb drivers have also been backported from the
   development (trunk) version.
   
 - the client development files have also received some care:
   the upsclient pkg-config file has been fixed, and the upsclient.h
   file allows older NUT clients to continue using the UPSCONN structure.
   
 For a complete list of changes, please refer to the ChangeLog file.

---------------------------------------------------------------------------
Release notes for NUT 2.2.0 - what's new since 2.0.5:

 - The new build infrastructure, using automake, is now used.
   This has major impact on the compilation and installation procedures,
   and thus on the NUT packaging.
   For more information, refer to UPGRADING and packaging/debian/ for
   an example of migration.

 - NUT now provides support for FreeDesktop Hardware Abstraction Layer

drivers/powerpanel.c
/*
 * powerpanel.c - Model specific routines for CyberPower text/binary
 *                protocol UPSes 
 *
 * Copyright (C)
 *	2007        Doug Reynolds <mav@wastegate.net>
 *	2007-2008   Arjen de Korte <adkorte-guest@alioth.debian.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 */

#include "main.h"
#include "serial.h"

#include "powerp-bin.h"
#include "powerp-txt.h"

static int	mode = 0;

static subdriver_t *subdriver[] = {
	&powpan_binary,
	&powpan_text,
	NULL
};

#define DRIVER_NAME	"CyberPower text/binary protocol UPS driver"
#define DRIVER_VERSION	"0.25"

/* driver description structure */
upsdrv_info_t upsdrv_info = {
	DRIVER_NAME,
	DRIVER_VERSION,
	"Doug Reynolds <mav@wastegate.net>\n" \
	"Arjen de Korte <adkorte-guest@alioth.debian.org>",
	DRV_EXPERIMENTAL,
	{ NULL }
};
/* FIXME: add a sub version for binary and text subdrivers? */

void upsdrv_initinfo(void)
{
	char	*s;

	dstate_setinfo("ups.mfr", "CyberPower");
	dstate_setinfo("ups.model", "[unknown]");
	dstate_setinfo("ups.serial", "[unknown]");

	subdriver[mode]->initinfo();

	/*
	 * Allow to override the following parameters
	 */
	if ((s = getval("manufacturer")) != NULL) {
		dstate_setinfo("ups.mfr", "%s", s);
	}
	if ((s = getval("model")) != NULL) {
		dstate_setinfo("ups.model", "%s", s);
	}
	if ((s = getval("serial")) != NULL) {
		dstate_setinfo("ups.serial", "%s", s);
	}

	upsh.instcmd = subdriver[mode]->instcmd;
	upsh.setvar = subdriver[mode]->setvar;
}

void upsdrv_updateinfo(void)
{
	static int	retry = 0;

	if (subdriver[mode]->updateinfo() < 0) {
		ser_comm_fail("Status read failed!");

		if (retry < 3) {
			retry++;
		} else {
			dstate_datastale();
		}

		return;
	}

	retry = 0;

	ser_comm_good();

	dstate_dataok();
}

void upsdrv_shutdown(void)
{
	int	i, ret;

	/*
	 * Try to shutdown with delay and automatic reboot if the power
	 * returned in the mean time (newer models support this).
	 */
	if (subdriver[mode]->instcmd("shutdown.return", NULL) == STAT_INSTCMD_HANDLED) {
		/* Shutdown successful */
		return;
	}

	/*
	 * Looks like an older type. Assume we're still on battery if
	 * we can't read status or it is telling us we're on battery.
	 */
	for (i = 0; i < MAXTRIES; i++) {

		ret = subdriver[mode]->updateinfo();
		if (ret >= 0) {
			break;
		}
	}

	if (ret) {
		/*
		 * When on battery, the 'shutdown.stayoff' command will make
		 * the UPS switch back on when the power returns.
		 */
		if (subdriver[mode]->instcmd("shutdown.stayoff", NULL) == STAT_INSTCMD_HANDLED) {
			upslogx(LOG_INFO, "Waiting for power to return...");
			return;
		}
	} else {
		/*
		 * Apparently, the power came back already, so we just need to reboot.
		 */
		if (subdriver[mode]->instcmd("shutdown.reboot", NULL) == STAT_INSTCMD_HANDLED) {
			upslogx(LOG_INFO, "Rebooting now...");
			return;
		}
	}

	upslogx(LOG_ERR, "Shutdown command failed!");
}

void upsdrv_initups(void)
{
	char	*version;

	version = getval("protocol");
	upsfd = ser_open(device_path);

	ser_set_rts(upsfd, 0);

	/*
	 * Try to autodetect which UPS is connected.
	 */
	for (mode = 0; subdriver[mode] != NULL; mode++) {

		if ((version != NULL) && strcasecmp(version, subdriver[mode]->version)) {
			continue;
		}

		ser_set_dtr(upsfd, 1);
		usleep(10000);

		if (subdriver[mode]->initups() > 0) {
			upslogx(LOG_INFO, "CyberPower UPS with %s protocol on %s detected", subdriver[mode]->version, device_path);
			return;
		}

		ser_set_dtr(upsfd, 0);
		usleep(10000);
	}

	fatalx(EXIT_FAILURE, "CyberPower UPS not found on %s", device_path);
}

void upsdrv_help(void)
{
}

void upsdrv_makevartable(void)
{
	addvar(VAR_VALUE, "ondelay", "Delay before UPS startup");
	addvar(VAR_VALUE, "offdelay", "Delay before UPS shutdown");

	addvar(VAR_VALUE, "manufacturer", "manufacturer");
	addvar(VAR_VALUE, "model", "modelname");
	addvar(VAR_VALUE, "serial", "serialnumber");
	addvar(VAR_VALUE, "protocol", "protocol to use [text|binary] (default: autodection)");
}

void upsdrv_cleanup(void)
{
	ser_set_dtr(upsfd, 0);
	ser_close(upsfd, device_path);
}
